File: Data Structures & Algorithms/house-robber/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n=len(nums)
        dp=[0]*n

        def helper(curr):
            
            if curr>=n:
                return 0
            if dp[curr]:
                return dp[curr]
            res=max(helper(curr+2),helper(curr+3))
            dp[curr]=res+nums[curr]
            return dp[curr]
        return max(helper(0),helper(1))

File: Data Structures & Algorithms/subsets/submission-4.java
class Solution {
    public void helper(int[] nums, int curr, List<Integer> res, List<List<Integer>> result) {
        result.add(new ArrayList<>(res));
        for (int i = curr; i < nums.length; i++) {
            res.add(nums[i]);            
            helper(nums, i + 1, res, result);
            res.remove(res.size() - 1);   
        }
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        helper(nums, 0, new ArrayList<>(), result);
        return result;
    }
}
